### ISSUE TITLE
perl 5.18 breaks AUTOLOADER inheritance

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#13456]
Migrated from [rt.perl.org#120694](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=120694) (status was 'resolved')

Searchable as RT120694$

**Comment by p5pRT** [comment]
### From @jandubois

#### Created by @jandubois



The Win32&#8203;::OLE module has a regression test \(t/3\_ole\.t\) that uses
AUTOLOAD inheritance\.
It has been working since at least 5\.006\, and possibly earlier up to 5\.16\.x\, but
is broken in 5\.18\.0\.  Here is a simplified test program&#8203;:

\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#
print bless\( \[\] => "foo" \)\->plugh\, "\\n";

package foo;
BEGIN \{ @&#8203;ISA = qw\(bar\) \}

sub AUTOLOAD \{
&nbsp;    my $self = shift;
&nbsp;    return if $AUTOLOAD eq 'DESTROY';
&nbsp;    print "foo&#8203;::AUTOLOAD&#8203;: $AUTOLOAD\\n";
&nbsp;    $AUTOLOAD = "SUPER&#8203;:" \. substr $AUTOLOAD\, rindex\($AUTOLOAD\, '&#8203;:'\);
&nbsp;    return $self\->$AUTOLOAD\(@&#8203;\_\);
\}

package bar;

sub AUTOLOAD \{
&nbsp;    print "bar&#8203;::AUTOLOAD&#8203;: $AUTOLOAD\\n";
&nbsp;    return "xyzzy";
\}

\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#

In perl versions prior to 5\.18\.0 it prints&#8203;:

$ perl ~/tmp/autoload\.pl
foo&#8203;::AUTOLOAD&#8203;: foo&#8203;::plugh
bar&#8203;::AUTOLOAD&#8203;: foo&#8203;::SUPER&#8203;::plugh
xyzzy
foo&#8203;::AUTOLOAD&#8203;: foo&#8203;::DESTROY
bar&#8203;::AUTOLOAD&#8203;: foo&#8203;::SUPER&#8203;::DESTROY


In 5\.18\.0 it recurses endlessly and puts garbage in $AUTOLOAD&#8203;:

$ /usr/local/ActivePerl\-5\.18/bin/perl ~/tmp/autoload\.pl
foo&#8203;::AUTOLOAD&#8203;: foo&#8203;::plugh
foo&#8203;::AUTOLOAD&#8203;: foo&#8203;::SUPER&#8203;::?
foo&#8203;::AUTOLOAD&#8203;: foo&#8203;::SUPER&#8203;::PER&#8203;::
foo&#8203;::AUTOLOAD&#8203;: foo&#8203;::SUPER&#8203;::
foo&#8203;::AUTOLOAD&#8203;: foo&#8203;::SUPER&#8203;::
foo&#8203;::AUTOLOAD&#8203;: foo&#8203;::SUPER&#8203;::
foo&#8203;::AUTOLOAD&#8203;: foo&#8203;::SUPER&#8203;::
foo&#8203;::AUTOLOAD&#8203;: foo&#8203;::SUPER&#8203;::
foo&#8203;::AUTOLOAD&#8203;: foo&#8203;::SUPER&#8203;::
foo&#8203;::AUTOLOAD&#8203;: foo&#8203;::SUPER&#8203;::

On Windows the garbage looks different&#8203;:

foo&#8203;::AUTOLOAD&#8203;: foo&#8203;::plugh
foo&#8203;::AUTOLOAD&#8203;: foo&#8203;::SUPER&#8203;::plugh
foo&#8203;::AUTOLOAD&#8203;: foo&#8203;::SUPER&#8203;::PER&#8203;::
foo&#8203;::AUTOLOAD&#8203;: foo&#8203;::SUPER&#8203;::
foo&#8203;::AUTOLOAD&#8203;: foo&#8203;::SUPER&#8203;::
foo&#8203;::AUTOLOAD&#8203;: foo&#8203;::SUPER&#8203;::
foo&#8203;::AUTOLOAD&#8203;: foo&#8203;::SUPER&#8203;::

<details><summary>Perl Info</summary>

```

Flags:
    category=core
    severity=low

Site configuration information for perl 5.18.1:

Configured by ActiveState at Fri Sep 20 20:12:23 PDT 2013.

Summary of my perl5 (revision 5 version 18 subversion 1) configuration:

  Platform:
    osname=darwin, osvers=10.8.0, archname=darwin-thread-multi-2level
    uname='darwin neko.activestate.com 10.8.0 darwin kernel version
10.8.0: tue jun 7 16:33:36 pdt 2011; root:xnu-1504.15.3~1release_i386
i386 '
    config_args='-ders -Dcc=gcc -Dusethreads -Duseithreads
-Uinstallusrbinperl -Ulocincpth= -Uloclibpth= -Duse64bitint -Ud_poll
-Ui_poll -Ui_libutil -Aldflags=-Wl,-search_paths_first
-Alddlflags=-Wl,-search_paths_first -Accflags=-arch x86_64
-Aldflags=-arch x86_64 -Alddlflags=-arch x86_64
-Accflags=-DUSE_SITECUSTOMIZE -Duselargefiles
-Accflags=-DPERL_RELOCATABLE_INCPUSH -Accflags=-fno-merge-constants
-Dprefix=/usr/local/ActivePerl-5.18
-Dprivlib=/usr/local/ActivePerl-5.18/lib
-Darchlib=/usr/local/ActivePerl-5.18/lib
-Dsiteprefix=/usr/local/ActivePerl-5.18/site
-Dsitelib=/usr/local/ActivePerl-5.18/site/lib
-Dsitearch=/usr/local/ActivePerl-5.18/site/lib -Dsed=/usr/bin/sed
-Duseshrplib -Dcf_by=ActiveState -Dcf_email=support@ActiveState.com'
    hint=recommended, useposix=true, d_sigaction=define
    useithreads=define, usemultiplicity=define
    useperlio=define, d_sfio=undef, uselargefiles=define, usesocks=undef
    use64bitint=define, use64bitall=define, uselongdouble=undef
    usemymalloc=n, bincompat5005=undef
  Compiler:
    cc='/usr/bin/gcc', ccflags ='-fno-common -DPERL_DARWIN
-no-cpp-precomp -arch x86_64 -DUSE_SITECUSTOMIZE
-DPERL_RELOCATABLE_INCPUSH -fno-merge-constants -fno-strict-aliasing
-pipe -fstack-protector',
    optimize='-O3',
    cppflags='-no-cpp-precomp -fno-common -DPERL_DARWIN
-no-cpp-precomp -arch x86_64 -DUSE_SITECUSTOMIZE
-DPERL_RELOCATABLE_INCPUSH -fno-merge-constants -fno-strict-aliasing
-pipe -fstack-protector'
    ccversion='', gccversion='Apple LLVM version 5.0 (clang-500.2.79)
(based on LLVM 3.3svn)', gccosandvers=''
    intsize=4, longsize=8, ptrsize=8, doublesize=8, byteorder=12345678
    d_longlong=define, longlongsize=8, d_longdbl=define, longdblsize=16
    ivtype='long', ivsize=8, nvtype='double', nvsize=8, Off_t='off_t',
lseeksize=8
    alignbytes=8, prototype=define
  Linker and Libraries:
    ld='/usr/bin/cc -mmacosx-version-min=10.5', ldflags ='
-Wl,-search_paths_first -arch x86_64 -fstack-protector'
    libpth=/usr/lib
    libs=-ldbm -ldl -lm -lutil -lc
    perllibs=-ldl -lm -lutil -lc
    libc=, so=dylib, useshrplib=true, libperl=libperl.dylib
    gnulibc_version=''
  Dynamic Linking:
    dlsrc=dl_dlopen.xs, dlext=bundle, d_dlsymun=undef, ccdlflags=' '
    cccdlflags=' ', lddlflags=' -bundle -undefined dynamic_lookup
-Wl,-search_paths_first -arch x86_64 -fstack-protector'

Locally applied patches:
    ACTIVEPERL_LOCAL_PATCHES_ENTRY


@INC for perl 5.18.1:
    /Users/jan/Library/ActivePerl-5.18/lib
    /usr/local/ActivePerl-5.18/site/lib
    /usr/local/ActivePerl-5.18/lib
    .


Environment for perl 5.18.1:
    DYLD_LIBRARY_PATH (unset)
    HOME=/Users/jan
    LANG=en_CA.UTF-8
    LANGUAGE (unset)
    LD_LIBRARY_PATH (unset)
    LOGDIR (unset)
    PATH=/usr/local/ActivePerl-5.18/bin:/usr/local/heroku/bin:/usr/local/var/rbenv/shims:/Users/jan/bin:/Users/jan/Dropbox/bin:/Users/jan/Library/ActivePerl-5.16/bin:/usr/local/ActivePerl-5.16/site/bin:/usr/local/ActivePerl-5.16/bin:/Users/jan/Library/Python/2.7/bin:/usr/local/bin:/usr/local/share/npm/bin:/usr/bin:/bin:/usr/sbin:/sbin:/usr/X11/bin
    PERL_BADLANG (unset)
    SHELL=/bin/bash

```
</details>

**Comment by p5pRT** [comment]
### From @jkeenan

On Wed Dec 04 14&#8203;:03&#8203;:25 2013\, jdb wrote&#8203;:
> This is a bug report for perl from jand@&#8203;activestate\.com\,
> generated with the help of perlbug 1\.39 running under perl 5\.18\.1\.
> 
> 
> \-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-
> \[Please describe your issue here\]
> 
> The Win32&#8203;::OLE module has a regression test \(t/3\_ole\.t\) that uses
> AUTOLOAD inheritance\.
> It has been working since at least 5\.006\, and possibly earlier up to
> 5\.16\.x\, but
> is broken in 5\.18\.0\.  Here is a simplified test program&#8203;:
> 
> \#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#
> print bless\( \[\] => "foo" \)\->plugh\, "\\n";
> 
> package foo;
> BEGIN \{ @&#8203;ISA = qw\(bar\) \}
> 
> sub AUTOLOAD \{
>     my $self = shift;
>     return if $AUTOLOAD eq 'DESTROY';
>     print "foo&#8203;::AUTOLOAD&#8203;: $AUTOLOAD\\n";
>     $AUTOLOAD = "SUPER&#8203;:" \. substr $AUTOLOAD\, rindex\($AUTOLOAD\, '&#8203;:'\);
>     return $self\->$AUTOLOAD\(@&#8203;\_\);
> \}
> 
> package bar;
> 
> sub AUTOLOAD \{
>     print "bar&#8203;::AUTOLOAD&#8203;: $AUTOLOAD\\n";
>     return "xyzzy";
> \}
> 
> \#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#
> 
> In perl versions prior to 5\.18\.0 it prints&#8203;:
> 
> $ perl ~/tmp/autoload\.pl
> foo&#8203;::AUTOLOAD&#8203;: foo&#8203;::plugh
> bar&#8203;::AUTOLOAD&#8203;: foo&#8203;::SUPER&#8203;::plugh
> xyzzy
> foo&#8203;::AUTOLOAD&#8203;: foo&#8203;::DESTROY
> bar&#8203;::AUTOLOAD&#8203;: foo&#8203;::SUPER&#8203;::DESTROY
> 
> 
> In 5\.18\.0 it recurses endlessly and puts garbage in $AUTOLOAD&#8203;:
> 
> $ /usr/local/ActivePerl\-5\.18/bin/perl ~/tmp/autoload\.pl
> foo&#8203;::AUTOLOAD&#8203;: foo&#8203;::plugh
> foo&#8203;::AUTOLOAD&#8203;: foo&#8203;::SUPER&#8203;::?
> foo&#8203;::AUTOLOAD&#8203;: foo&#8203;::SUPER&#8203;::PER&#8203;::
> foo&#8203;::AUTOLOAD&#8203;: foo&#8203;::SUPER&#8203;::
> foo&#8203;::AUTOLOAD&#8203;: foo&#8203;::SUPER&#8203;::
> foo&#8203;::AUTOLOAD&#8203;: foo&#8203;::SUPER&#8203;::
> foo&#8203;::AUTOLOAD&#8203;: foo&#8203;::SUPER&#8203;::
> foo&#8203;::AUTOLOAD&#8203;: foo&#8203;::SUPER&#8203;::
> foo&#8203;::AUTOLOAD&#8203;: foo&#8203;::SUPER&#8203;::
> foo&#8203;::AUTOLOAD&#8203;: foo&#8203;::SUPER&#8203;::
> 

Confirmed on blead on the dromedary server \(linux x86\_64\)\.

> On Windows the garbage looks different&#8203;:
> 
> foo&#8203;::AUTOLOAD&#8203;: foo&#8203;::plugh
> foo&#8203;::AUTOLOAD&#8203;: foo&#8203;::SUPER&#8203;::plugh
> foo&#8203;::AUTOLOAD&#8203;: foo&#8203;::SUPER&#8203;::PER&#8203;::
> foo&#8203;::AUTOLOAD&#8203;: foo&#8203;::SUPER&#8203;::
> foo&#8203;::AUTOLOAD&#8203;: foo&#8203;::SUPER&#8203;::
> foo&#8203;::AUTOLOAD&#8203;: foo&#8203;::SUPER&#8203;::
> foo&#8203;::AUTOLOAD&#8203;: foo&#8203;::SUPER&#8203;::
> 
> 
>

**Comment by p5pRT** [comment]
The RT System itself - Status changed from 'new' to 'open'

**Comment by p5pRT** [comment]
### From @andk

"James E Keenan via RT" \<perlbug\-followup@&#8203;perl\.org> writes&#8203;:

> Confirmed on blead on the dromedary server \(linux x86\_64\)\.

Bisected to v5\.17\.3\-467\-gaae4380

&nbsp;  commit aae438050a206a8067e068319ab3ee2348009463
&nbsp;  Author&#8203;: Father Chrysostomos \<sprout@&#8203;cpan\.org>
&nbsp;  Date&#8203;:   Mon Sep 17 16&#8203;:24&#8203;:40 2012 \-0700

&nbsp;      \[perl \#114924\] Make method calls work with :&#8203;:SUPER packages


\-\- 
andreas

**Comment by p5pRT** [comment]
### From @cpansprout

FIxed in 257dc59\.

\-\- 

Father Chrysostomos

**Comment by p5pRT** [comment]
@cpansprout - Status changed from 'open' to 'resolved'

**Comment by p5pRT** [comment]
### From @rjbs

\* Father Chrysostomos via RT \<perlbug\-followup@&#8203;perl\.org> \[2013\-12\-05T09&#8203;:16&#8203;:13\]
> FIxed in 257dc59\.

Suggested for 5\.18\.2?

\-\- 
rjbs

**Comment by p5pRT** [comment]
### From perl5-porters@perl.org

Ricardo Signes wrote&#8203;:
> \* Father Chrysostomos via RT \<perlbug\-followup@&#8203;perl\.org> \[2013\-12\-05T09&#8203;:16&#8203;:13\]
> > FIxed in 257dc59\.
>
> Suggested for 5\.18\.2?

Yes\, that's a good idea\.
